### PR TITLE
QE-7173 performance testing for reporting-api-ingest

### PR DESCRIPTION
**Pull request by zmknox**

### CODE REVIEW

**Review by amalhotragannett — COMMENTED** [review]
Looking good from the initial look. Can you please add a readme so I can test it locally 👍

**Comment by zmknox** [comment]
@amalhotragannett I had one unsaved based on the elections readme. I added more information on the config files I made and pushed it up.

File: reporting-api-performance-tests/package.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  "author": {
+    "name": "Quality Engineering",
+    "email": "QualityEngineering@gannett.com"
+  },
+  "main": "src/index.js",
+  "scripts": {
+    "generate": "node src/index.js --cleanTestDirectory",
+    "setup:faas-artillery": "vault read -field=key secret/quality-engineering/gcp/production/service-accounts/api-services-faas-artillery > keys/service-account.json",
+    "setup": "npm run setup:faas-artillery",
+    "test": "docker run --rm -v $(pwd)/keys:/keys -v $(pwd)/test:/test -e GOOGLE_APPLICATION_CREDENTIALS=/keys/service-account.json -e quay.io/gannett/faas-artillery:latest run --dir /test --config /test/reporting-api-ingest-performance-test-central-config.yaml",
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/package.json:14]: type right before the quay image.

File: reporting-api-performance-tests/src/artillery-utils.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    scenerioEntries.forEach((scenerio) => {
+      for (const endpoint of endpoints) {
+        yamlConfig.scenarios.push(artilleryUtils._getScenerioConfig(scenerio, endpoint));
+      }
+    });
+
+    const doc = new YAML.Document();
+    doc.contents = yamlConfig;
+    fs.writeFileSync(path.resolve(__dirname, '../test', `reporting-api-ingest-performenace-tests.yaml`), doc.toString());
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/artillery-utils.js:51]:
```suggestion
    fs.writeFileSync(path.resolve(__dirname, '../test', `reporting-api-ingest-performance-tests.yaml`), doc.toString());
```

File: reporting-api-performance-tests/src/artillery-utils.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    
+
+    const faasArtilleryConfig = {
+      name: `reporting-api-ingest-performenace-tests`,
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/artillery-utils.js:55]:
```suggestion
      name: `reporting-api-ingest-performance-tests`,
```

File: reporting-api-performance-tests/src/artillery-utils.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      script: `reporting-api-ingest-performenace-tests.yaml`,
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/artillery-utils.js:56]:
```suggestion
      script: `reporting-api-ingest-performance-tests.yaml`,
```

File: reporting-api-performance-tests/src/artillery-utils.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      workers: 1,
+      rounds: 1
+    };
+    doc.contents = faasArtilleryConfig;
+    fs.writeFileSync(path.resolve(__dirname, '../test', `reporting-api-ingest-performenace-tests-config.yaml`), doc.toString());
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/artillery-utils.js:61]:
```suggestion
    fs.writeFileSync(path.resolve(__dirname, '../test', `reporting-api-ingest-performance-tests-config.yaml`), doc.toString());
```

File: reporting-api-performance-tests/package.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  "author": {
+    "name": "Quality Engineering",
+    "email": "QualityEngineering@gannett.com"
+  },
+  "main": "src/index.js",
+  "scripts": {
+    "generate": "node src/index.js --cleanTestDirectory",
+    "setup:faas-artillery": "vault read -field=key secret/quality-engineering/gcp/production/service-accounts/api-services-faas-artillery > keys/service-account.json",
+    "setup": "npm run setup:faas-artillery",
+    "test": "docker run --rm -v $(pwd)/keys:/keys -v $(pwd)/test:/test -e GOOGLE_APPLICATION_CREDENTIALS=/keys/service-account.json quay.io/gannett/faas-artillery:latest run --dir /test --config /test/reporting-api-ingest-performance-test-central-config.yaml",
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/package.json:14]:
```suggestion
    "test": "docker run --rm -v $(pwd)/keys:/keys -v $(pwd)/test:/test -e GOOGLE_APPLICATION_CREDENTIALS=/keys/service-account.json quay.io/gannett/faas-artillery:latest run --dir /test --config /test/reporting-api-ingest-performance-tests-config.yaml",
```

**Review by amalhotragannett — COMMENTED** [review]
The generated test is not compatible with artillery. It is throwing some errors. 

`artillery run ./test/reporting-api-ingest-performance-tests.yaml`

**Review by amalhotragannett — COMMENTED** [review]
Everything is looking good. Just one scenario is throwing a `400` response status. Is this an intended behavior? 

```
  - name: USAT network error (east)
    flow:
      - post:
          url: https://reports.east.gannettinnovation.com/
          headers:
            User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/605.1.15
              (KHTML, like Gecko) Version/14.0 Safari/605.1.15
          body: '"[{\"age\":0,\"body\":{\"elapsed_time\":531,\"method\":\"POST\",\"phase\":\"application\",\"protocol\":\"h3-Q050\",\"referrer\":\"https://usatoday.com/opinion\",\"sampling_fraction\":1,\"server_ip\":\"205.43.241.22\",\"status_code\":400,\"type\":\"ok\"},\"type\":\"network-error\",\"url\":\"https://usatoday.com/\"}]"'
```

**Comment by zmknox** [comment]
@amalhotragannett Looks like there was a double stringify, fixed

File: reporting-api-performance-tests/src/config/endpoints.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "name": "central",
+    "target": "https://reports.central.gannettinnovation.com"
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/config/endpoints.json:4]: Please change the url to dev instance.

File: reporting-api-performance-tests/src/config/reports.json [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          "sampling_fraction": 1,
+          "server_ip": "205.43.241.22",
+          "status_code": 400,
+          "type": "ok"
+        },
+        "type": "network-error",
+        "url": "https://usatoday.com/"
+      }
+    ],
+    "userAgemt": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Safari/605.1.15"
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/config/reports.json:22]: Typo

File: reporting-api-performance-tests/src/config/reports.json [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+          "sampling_fraction": 1,
+          "server_ip": "2020.10.12.13.0.1",
+          "status_code": 304,
+          "type": "ok"
+        },
+        "type": "network-error",
+        "url": "https://example.com/"
+      }
+    ],
+    "userAgemt": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36"
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/config/reports.json:44]: typo userAgent

File: reporting-api-performance-tests/src/config/reports.json [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+          "phase": "application",
+          "protocol": "h3-Q050",
+          "referrer": "https://example.com/referrer",
+          "sampling_fraction": 1,
+          "server_ip": "2020.10.12.13.0.1",
+          "status_code": 304,
+          "type": "ok"
+        },
+        "type": "network-error",
+        "url": "https://example.com/"
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/config/reports.json:41]:
```suggestion
        "url": "https://example.com/",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36"
```

File: reporting-api-performance-tests/src/config/reports.json [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+          "phase": "application",
+          "protocol": "h3-Q050",
+          "referrer": "https://usatoday.com/opinion",
+          "sampling_fraction": 1,
+          "server_ip": "205.43.241.22",
+          "status_code": 400,
+          "type": "ok"
+        },
+        "type": "network-error",
+        "url": "https://usatoday.com/"
```
Comment (amalhotragannett) [inline reporting-api-performance-tests/src/config/reports.json:19]:
```suggestion
        "url": "https://usatoday.com/",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Safari/605.1.15"
```

Comment (zmknox) [reply reporting-api-performance-tests/src/config/reports.json:19]: I was keeping user agent outside of that "body" because its a header

Comment (amalhotragannett) [reply reporting-api-performance-tests/src/config/reports.json:19]: It's a header value but also required in the request body.

Comment (amalhotragannett) [reply reporting-api-performance-tests/src/config/reports.json:19]: The header is used by the cloud run to give attributes to the dataflow job. Where the body is attached as is inside of BigQuery without any modifications.

**Review by amalhotragannett — APPROVED** [review]
Good work!!! Thanks for getting this setup 👍

**Pull request merged**